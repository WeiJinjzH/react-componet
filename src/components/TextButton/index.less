.text-button {
    background: none;
    border: none;
    color: #1890ff;
    cursor: pointer;
    outline: none;
    white-space: nowrap;

    &:hover {
        color: #4ba7fd;
    }
    &:active {
        color: #1890ff;
        text-decoration: none;
    }
    &:disabled {
        color: #ccc;
        cursor: not-allowed;
    }
}

.text-button--link {
    &:focus {
        text-decoration: none;
    }
    &:hover {
        color: #4ba7fd;
        text-decoration: none;
    }
    &:active {
        color: #1890ff;
        text-decoration: none;
    }
}

.text-button--disabled {
    color: #ccc !important;
    cursor: not-allowed;
}