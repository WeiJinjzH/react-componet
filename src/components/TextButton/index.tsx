import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import utils from 'src/utils'
import './index.less'

interface TextButtonProps {
    /** 是否可小窗口预览, 默认值 true */
    canPreview?: boolean;
    download?: any;
    href?: string;
    className?: string;
    title?: string;
    style?: React.CSSProperties;
    type?: 'button'|'submit'|'reset';
    disabled?: boolean;
    onClick?: React.MouseEventHandler
}

function handlePreview(href: string) {
    if (!utils.showPreviewModal({ href })) {
        window.open(href)
    }
}

class TextButton extends PureComponent<TextButtonProps> {
    static propTypes = {
        type: PropTypes.oneOf(['button', 'submit', 'reset']),
        canPreview: PropTypes.bool,
    }

    static defaultProps = {
        type: 'button',
        canPreview: true,
    }

    constructor(props: TextButtonProps) {
        super(props)
        if (props.href && props.onClick) {
            window.console.warn('TextButton: "href" 与 "onClick" 同时存在时, "onClick" 将不生效')
        }
    }

    render() {
        const {
            download, href, className, style, title, disabled, onClick, type, children, canPreview,
        } = this.props
        if (href) { // 当'href'为真值时, 元素渲染为<a>, 否则渲染为<button>
            return (
                <a
                    className={classNames({
                        'text-button': true,
                        'text-button-link': true,
                        'text-button-link--disabled': disabled,
                        [className]: !!className,
                    })}
                    onClick={(e) => {
                        if (!disabled && canPreview) {
                            handlePreview(href)
                        }
                        if (disabled || canPreview) {
                            e.preventDefault()
                            e.stopPropagation()
                        }
                    }}
                    style={style}
                    title={title}
                    href={href}
                    download={download}
                    target="_blank"
                    rel="noopener noreferrer"
                >
                    {this.props.children}
                </a>
            )
        }
        return (
            <button
                className={classNames({
                    'text-button': true,
                    [className]: !!className,
                })}
                style={style}
                title={title}
                disabled={disabled}
                onClick={onClick}
                type={type}
            >
                {children || ''}
            </button>
        )
    }
}

export default TextButton
