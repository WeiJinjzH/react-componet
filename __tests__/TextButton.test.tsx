import { render, shallow, mount } from 'enzyme'
import React from 'react'
import TextButton from 'src/components/TextButton'

describe('Component: TextButton', () => {
    it('render test', () => {
        const wrapper = shallow(<TextButton />)
        expect(wrapper.find('.text-button')).toHaveLength(1)
        expect(wrapper.html()).toContain('<button')
        expect(wrapper).toMatchSnapshot()

        const wrapper2 = shallow(<TextButton>Text</TextButton>)
        expect(wrapper2.hasClass('text-button')).toBeTruthy()
        expect(wrapper2).toMatchSnapshot()

        const wrapper3 = shallow(<TextButton disabled>Text</TextButton>)
        expect(wrapper3).toMatchSnapshot()

        const wrapper4 = shallow(<TextButton canPreview={false} />)
        expect(wrapper4.find('.text-button')).toHaveLength(1)
        expect(wrapper4).toMatchSnapshot()

        const wrapper5 = shallow((<TextButton href="/" disabled />))
        expect(wrapper5.html()).toContain('<a')
        expect(wrapper5.hasClass('text-button-link')).toBeTruthy()
        expect(wrapper5.hasClass('text-button-link--disabled')).toBeTruthy()
    })


    it('interaction test', () => {
        const onClick = jest.fn()
        const wrapper = mount((<TextButton href="/" onClick={onClick} disabled />))
        expect(wrapper).toMatchSnapshot()
        wrapper.find('.text-button').simulate('click')
        expect(onClick).toBeCalledTimes(0)

        /** href 与 onClick 同时存在时, onClick将不生效 */
        wrapper.setProps({ disabled: false })
        wrapper.find('.text-button').simulate('click')
        expect(onClick).toBeCalledTimes(0)

        wrapper.setProps({ href: undefined })
        expect(wrapper).toMatchSnapshot()
        wrapper.find('.text-button').simulate('click')
        expect(onClick).toBeCalledTimes(1)

        wrapper.setProps({ href: '.png' })
        wrapper.find('.text-button').simulate('click')
        expect(onClick).toBeCalledTimes(1)

        wrapper.setProps({ href: '.pdf' })
        wrapper.find('.text-button').simulate('click')
        expect(onClick).toBeCalledTimes(1)

        wrapper.setProps({ canPreview: false })
        wrapper.find('.text-button').simulate('click')
        expect(onClick).toBeCalledTimes(1)
    })
})
