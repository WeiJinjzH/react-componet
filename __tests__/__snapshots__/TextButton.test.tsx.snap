// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TextButton check: disabled 1`] = `
<TextButton
  canPreview={true}
  disabled={true}
  href="/"
  onClick={[MockFunction]}
  type="button"
>
  <a
    className="text-button text-button-link text-button-link--disabled"
    href="/"
    onClick={[Function]}
    rel="noopener noreferrer"
    target="_blank"
  />
</TextButton>
`;

exports[`Component: TextButton check: disabled 2`] = `
<TextButton
  canPreview={true}
  disabled={false}
  onClick={[MockFunction]}
  type="button"
>
  <button
    className="text-button"
    disabled={false}
    onClick={[MockFunction]}
    type="button"
  />
</TextButton>
`;

exports[`Component: TextButton render test 1`] = `
<button
  className="text-button"
  type="button"
/>
`;

exports[`Component: TextButton render test 2`] = `
<button
  className="text-button"
  type="button"
>
  Text
</button>
`;

exports[`Component: TextButton render test 3`] = `
<button
  className="text-button"
  disabled={true}
  type="button"
>
  Text
</button>
`;

exports[`Component: TextButton render test 4`] = `
<button
  className="text-button"
  type="button"
/>
`;
